release: Nvim 0.5.1

justinmk opened the pull request:
Please test this locally:

```
VIMRUNTIME=/path/to/neovim/runtime /path/to/neovim/build/bin/nvim
```

## Notes

- 14cdaca6a8180c6e3079cd2e73fa75a5eb5253da fixes a problem not noticed on master because of the vim.lsp.diagnostic => vim.diagnostic move
    - also fixes some tests that accidentally depended on the previous session

mjlbach committed:
fix(lsp): avoid ipairs on non-sequential tables (#15059)

ipairs terminates on the first nil index when iterating over table keys:

for i,k in ipairs( {[1] = 'test', [3] = 'test'} ) do
  print(i, k)
end

prints:
1 test

Instead, use pairs which continues iterating over the entire table:

for i,k in pairs( {[1] = 'test', [3] = 'test'} ) do
  print(i, k)
end

prints:
1 test
3 test

simrat39 committed:
lsp(start_client): Allow passing custom workspaceFolders to the LSP (#15132)

Some language servers *cough*rust-analyzer*cough* need an empty/custom
workspaceFolders for certain usecases. For example, rust-analyzer
needs an empty workspaceFolders table for standalone file support
(See https://github.com/rust-analyzer/rust-analyzer/pull/8955).

This can also be useful for other languages that need to commonly
open a certain directory (like flutter or lua), which would help
prevent spinning up a new language server altogether.

In case no workspaceFolders are passed, we fallback to what we had
before.

mfussenegger committed:
feat(lsp): allow root_dir to be nil (#15430)

According to the protocol definition `rootPath`, `rootUri` and
`workspaceFolders` are allowed to be null.

Some language servers utilize this to provide "single file" support.
If all three are null, they don't attempt to index a directory but
instead only provide capabilities for a single file.

Commit:
fix(lsp): check if buffer is valid in changetracking (#15505)

mjlbach committed:
feat(lsp)!: change handler signature #15504

Commit:
fix(lsp): update workspace/applyEdit handler signature (#15573)

mfussenegger committed:
fix(lsp): adapt codelens resolve to handler signature change (#15578)

Follow up to https://github.com/neovim/neovim/pull/15504

mjlbach committed:
feat(lsp): improve logging (#15636)

* Simplify rpc encode/decode messages to rpc.send/rcp.receive
* Make missing handlers message throw a warning
* Clean up formatting style in log
* Move all non-RPC loop messages to trace instead of debug
* Add format func option to log to allow newlines in per log entry

hrsh7th committed:
feat(lsp): improve vim.lsp.util.apply_text_edits (#15561)

- Fix the cursor position after applying TextEdits
- Support reversed range of TextEdit
- Invoke nvim_buf_set_text one by one

mfussenegger committed:
fix(lsp): update lsp-handler signature in call_hierarchy (#15738)

This fixes the handler signature and also prevents n+1 requests firing
if there are multiple clients.

(The first `prepareCallHierarchy` handler is called once per client,
each invocation used `buf_request` to make more requests using *all*
clients)

justinmk committed:
NVIM v0.5.1

BREAKING CHANGES:
d83df7f7b5c1 feat(lua)!: register_keystroke_callback => on_key
cd8f6c5fb785 feat(lsp)!: change handler signature #15504

FEATURES:
915dda3f96b3 feat(jobstart): add parameter to close stdin

FIXES:
f8e0011534a3 #15732 fix(inccommand): ignore trailing commands only for *previewed* command
2132c063af3a backport: fix(windowing): positioning of relative floats
51d6b26729f4 #15495 backport: tests(lua/on_yank): assert conditions that fail correctly
f7002337c02a #15482 backport: fix(lua): verify buffer in highlight.on_yank
6bda2f56eb01 #15454 backport: fix(window.c): win_close from other tabpage
be58ba250ee2 #15372 backport: fix(autocmd.c): fix conditions in block_autocmds, unblock_autocmds
d0e9a11e395a backport: refactor(sign): include longer sign column option
5c42376c15e2 backport: fix(sign): reset auto sign column with minimum in float win minimal style
41f761130e4f backport: fix(decorations): crash when :bdelete (extmark_free_all) after clear_namespace
cf62554e5aea #15111 backport: fix(:source): copy curbuf lines to memory before sourcing
6436100b6e4b #14809 backport: fix(:source, nvim_exec): handle Vimscript line continuations
917f30666657 #15043 backport: test/memory_usage_spec: skip on MacOS
a9cca1b050f4 #14984 backport: fixup(clipboard): Fix error not properly handled
ae89330ec047 #14982 backport: fix(vim.opt): vimL map string values not trimmed
2229e99ef918 #14962 backport: fixup(clipboard): Use case matching
b6b12ea7c3f2 #15489 fix(man.vim): filetype=man is too eager
6f965f41dfd6 build: use RelWithDebInfo build for nightlies, Release for releases
f027c5e1e463 build: update appdata.xml version in release commit
8336488ce1cd test(treesitter): skip all parsers tests if parsers aren't installed
008b83f5a28e Rename stdin to stdin_mode (fixes Windows build)

FIXES (LSP):
132053c1d2bd #15523 backport: fix(lsp): resolve bufnr in buf_is_attached
a265201307f6 backport: fix(lsp): Ensure human readable errors are printed
33000bd9cff3 backport: fix(lsp): Ensure users get feedback on references/symbols errors or empty results
9f73b7c214c6 #14954 backport: fix(lsp): correctly check for windows in lsp logger
eaa1c4737792 #15023 backport: fix(lsp): restore diagnostics extmarks that were moved to the last edit line
989ccb822203 #15011 backport: fix(lsp): restore diagnostics extmarks on buffer changes
2ae4c96d9176 backport: fix(lsp): prevent double <text> for cached plaintext markup
7b0ae589f0a7 feat(lsp): allow root_dir to be nil (#15430) (Mathias Fußenegger)
8ec5bc91262a lsp(start_client): Allow passing custom workspaceFolders to the LSP (#15132) (sim)
959cf5e53c79 fix(lsp): check if buffer is valid in changetracking (#15505) (Jose Alvarez)
dc15b3a92c9b fix(lsp): avoid ipairs on non-sequential tables (#15059) (Michael Lingelbach)
18375c6df609 feat(lsp): improve vim.lsp.util.apply_text_edits (#15561) (hrsh7th)
7b1315fe6113 feat(lsp): improve logging (#15636) (Michael Lingelbach)

justinmk committed:
fix(release.sh): ignore failed "rm"

justinmk committed:
version bump

mjlbach commented on CMakeLists.txt:
```diff
@@ -137,8 +137,8 @@ set_property(CACHE CMAKE_BUILD_TYPE PROPERTY
 set(NVIM_VERSION_MINOR 5)
-set(NVIM_VERSION_PATCH 0)
-set(NVIM_VERSION_PRERELEASE "") # for package maintainers
+set(NVIM_VERSION_PATCH 2)
```
Minor question, why is this 2?

clason replied:
I think this is for _after_ the release? (So a build from `release-0.5` reports itself as `0.5.2-dev` -- see the following line.)

justinmk replied:
it should have been 1 before.

mjlbach replied:
Got it, thanks!

bfredl replied:
because I dun goofed

justinmk commented:
`CI / linux functionaltest-lua (cc=gcc)` for reference:

```
021-09-26T16:05:44.4420590Z [32m[2m[----------][0m[0m Running tests from [36mtest/functional/plugin/lsp_spec.lua[0m
...
2021-09-26T16:05:44.8291539Z [32m[2m[ RUN      ][0m[0m LSP basic_init test should succeed with manual shutdown: [2m50.58 ms[0m [1m[32mOK[0m[0m
2021-09-26T16:23:20.4923045Z -- Tests exited non-zero: Process terminated due to timeout
2021-09-26T16:23:20.4923982Z CMake Error at /home/runner/work/neovim/neovim/cmake/RunTests.cmake:90 (message):
2021-09-26T16:23:20.4924843Z   functional tests failed with error: Process terminated due to timeout
```

justinmk commented:
this test is timing out:

> LSP basic_init test should check the body sent attaching before init

justinmk committed:
fix(lsp): fix handler signature, tests

- not necessary on master: got lost in the vim.lsp.diagnostic => vim.diagnostic migration
- fix tests which accidentally depended on previous session
- ref #15504
